File: 2023/kotlin/src/Day06.kt
private fun countRaceWins(recordTime: Long, recordDistance: Long) =
    (0..recordTime).count { distance -> distance * (recordTime - distance) > recordDistance }

private fun part1(input: String): Int {
    val timesAndDistances = input.lines().map { it.longs() }
    val races = timesAndDistances[0] zip timesAndDistances[1]

    return races
        .map { (time, distance) -> countRaceWins(time, distance) }
        .reduce(Int::times)
}

private fun part2(input: String): Int {
    val timeAndDistance = input.lines().map { it.replace(" ", "").longs()[0] }

    return countRaceWins(timeAndDistance[0], timeAndDistance[1])
}

fun main() {
    val input = readInput("Day06")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day06.swift
struct Day06: AdventDay {
  var input: String

  func countRaceWins(recordTime: Int, recordDistance: Int) -> Int {
    (0...recordTime)
      .filter({ distance in
        distance * (recordTime - distance) > recordDistance
      })
      .count
  }

  func part1() -> Any {
    let timesAndDistances = input.lines().map { $0.ints() }
    let races = zip(timesAndDistances[0], timesAndDistances[1])

    return
      races
      .map { timeAndDistance in
        let (time, distance) = timeAndDistance
        return countRaceWins(recordTime: time, recordDistance: distance)
      }
      .reduce(1, *)
  }

  func part2() -> Any {
    let timeAndDistance = input.lines().map { $0.replacingOccurrences(of: " ", with: "").ints()[0] }

    return countRaceWins(recordTime: timeAndDistance[0], recordDistance: timeAndDistance[1])
  }
}
